Add Google Places autocomplete

klappradla opened the pull request:
This adds a new SimpleForm input type called *"city"* which enhances a string input with Google Places powered autocomplete for Cities as explained in #366 .

Loading the required libs from Google and then wiring up an autocomplete field requires an API key. I added one for development *(limited to `http://localhost:300/`)* and will add some env variables for staging and production. I cannot really estimate if we'll hit the rate limit for the free plan on my account. Let's just give it a try 🙈 

If you like, I can also add the lon/lat thing @carpodaster (it's super straight forward) but I would probably then add a `City` model to not make the application form more complex than it already is. Would also do this in a separate PR then I guess. Should I?

carpodaster commented on config/initializers/google_maps.rb:
```diff
@@ -0,0 +1,2 @@
+# Set Google Maps API key for development
+ENV['GOOGLE_MAPS_API_KEY'] ||= 'AIzaSyAPJUuLWQVCWoTIB2UkPAfgQSmPkDho9H0'
```
Could you also add this to `.env-sample` ?

klappradla replied:
Oh, totally overlooked the app is using Foreman. I'll add it there, but would also keep the initializer (e.g. for test env).

carpodaster commented on app/inputs/city_input.rb:
```diff
@@ -0,0 +1,5 @@
+class CityInput < SimpleForm::Inputs::StringInput
```
:heart:

I don't work with simple form much and thus I have never done custom inputs. While working on the character limited input the other day, I went that route, too. Never seen custom simple form inputs in years, now twice in two days :wink:

carpodaster commented:
As for the lat/lng and a `City` model: I'm not sure if that would make the form less complex. IIRC, you get the lat/lng from the API repsonse; couldn't we just populate two hidden form fields (or one composite `Array` type field) that gets stored under the hood?

klappradla commented:
Yeah, sounds fine. I'll add the lat/lng thing and update this PR 👍

klappradla commented:
Also: we can "lift" the rate limit from 1000 reg per day to **150000** *(which is probably enough)* by ["validating our identity"](https://developers.google.com/places/web-service/usage?hl=de#enable_billing) with a credit card. This does not mean, the credit card will be billed when exceeding the rate limit. Same as at Heroku, it's just for validating yourself.

I can use my credit card to do this with the current keys. But for the sake of simplicity, I think it would be better if a person with access to the credit card used for Heroku, etc. would do this, to not have these types of credentials spread over too many people (cc @alicetragedy)

alicetragedy commented:
@klappradla sure thing, I'd suggest that we use the Foundation CC that @anikalindtner has access to for this. Possibly, since we were thinking of creating a Google account for everything RGSoC-related anyway as once suggested by Ana (to keep all docs, etc... in the same place) we could also generate new API keys then, if you want (doesn't have to be though).

klappradla commented:
Sounds like a plan. Just ping me once you're doing it and I can tell you which services to pick, which origins to whitelist, etc.

klappradla commented:
Added the lat/lng stuff. I also moved the google api key into a global JS var - seemed more env-var-like to me...
Getting my Javascript code to pass with jshint was super hard 🙈  *(semicolons hell)*

F3PiX commented:
👏  👏

carpodaster commented on README.md:
````diff
@@ -114,8 +114,25 @@ directory the variables from `.env` will be loaded into the environment.
+1. Run the `foreman` command before any command
+  ```sh
+  # e.g.
+  foreman run rails server
````
I wasn't aware that foreman honours `.env`. For some reason I always thought we were using [dotenv](https://github.com/bkeepers/dotenv) 👍

carpodaster commented on spec/inputs/city_input_spec.rb:
```diff
@@ -0,0 +1,29 @@
+describe CityInput do
+  include RSpec::Rails::HelperExampleGroup
+
+  describe '#input' do
```
The test setup looks intimidating! I will copy/paste that for my own custom input ☺️  Thanks!

carpodaster commented on .env-example:
```diff
@@ -4,3 +4,6 @@ MAILTRAP_PASSWORD=InboxPassword
 CAMO_KEY=supersecret
+
+# Google Maps API key for e.g. city autocomplete
+GOOGLE_MAPS_API_KEY=YourGoogleMapsAPIKey
```
Shall we add the one you created for `localhost:3000` here? Otherwise it would override the setting in `config/google_maps.rb`.

klappradla replied:
We can add it, but I thought overwriting was the purpose here anyways.

carpodaster replied:
You mean when someone copies the `.env-sample`, they want to add new keys anyway? Makes sense. I was thinking that Google Maps not working anymore could be a hidden side-effect in the event that people just want to use, say, Mailgun and copied that `.env` for that end.

But let's stick with it as it is right now!
